src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/MapCode/MapCodeEndpoint.cs

```diff
@@ -0,0 +1,557 @@
+        LSP.MapCodeParams request,
+        RazorRequestContext context,
+        CancellationToken cancellationToken)
+    {
+        // TO-DO: Apply updates to the workspace before doing mapping. This is currently
+        // unimplemented by the client, so we won't bother doing anything for now until
+        // we determine what kinds of updates the client will actually send us.
+        if (request.Updates is not null)
+        {
+            return null;
+        }
+
+        using var _ = ArrayBuilderPool<TextDocumentEdit>.GetPooledObject(out var changes);
+        foreach (var mapping in request.Mappings)
+        {
+            if (mapping.TextDocument is null || mapping.FocusLocations is null)
+            {
+                continue;
+            }
+
+            var documentContext = _documentContextFactory.TryCreateForOpenDocument(mapping.TextDocument.Uri);
+            if (documentContext is null)
+            {
+                continue;
+            }
+
+            var (projectEngine, importSources) = await InitializeProjectEngineAsync(documentContext.Snapshot).ConfigureAwait(false);
+            var tagHelperContext = await documentContext.GetTagHelperContextAsync(cancellationToken).ConfigureAwait(false);
+            var fileKind = FileKinds.GetFileKindFromFilePath(documentContext.FilePath);
+            var extension = Path.GetExtension(documentContext.FilePath);
+
+            foreach (var content in mapping.Contents)
+            {
+                if (content is null)
+                {
+                    continue;
+                }
+
+                // We create a new Razor file based on each mapping's content in order to get the syntax tree that we'll later use to map.
+                var sourceDocument = RazorSourceDocument.Create(content, "Test" + extension);
+                var codeToMap = projectEngine.ProcessDesignTime(sourceDocument, fileKind, importSources, tagHelperContext.TagHelpers);
+
+                await MapCodeAsync(codeToMap, mapping.FocusLocations, changes, documentContext, cancellationToken).ConfigureAwait(false);
+            }
+        }
+
+        var workspaceEdits = new LSP.WorkspaceEdit
+        {
+            DocumentChanges = changes.ToArray()
+        };
+
+        return workspaceEdits;
+    }
+
+    private async Task MapCodeAsync(
+        RazorCodeDocument codeToMap,
+        LSP.Location[][] locations,
+        ImmutableArray<TextDocumentEdit>.Builder changes,
+        VersionedDocumentContext documentContext,
+        CancellationToken cancellationToken)
+    {
+        var syntaxTree = codeToMap.GetSyntaxTree();
+        if (syntaxTree is null)
+        {
+            return;
+        }
+
+        var nodesToMap = ExtractValidNodesToMap(syntaxTree.Root);
+        if (nodesToMap.Count == 0)
+        {
+            return;
+        }
+
+        await MapCodeAsync(locations, [.. nodesToMap], changes, documentContext, cancellationToken).ConfigureAwait(false);
```

Comment:
Seems like copying `nodesToMap` into an array is unnecessary, can just pass in the list